File: 20. Valid Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        
        parentheses_stack = []
        close_dict = {')': '(', ']': '[', '}': '{'}

        for i in s: 
            if i in close_dict:
                if parentheses_stack and close_dict[i] == parentheses_stack[-1]:
                    parentheses_stack.pop()
                else:
                    return False
            else:
                parentheses_stack.append(i)

        return not parentheses_stack

File: C-Sharp Solutions/Stacks/20. Valid Parentheses.cs
public class Solution {
    public bool IsValid(string s) {
        Dictionary<char, char> dict_keys = new Dictionary<char, char>{
            {'(', ')'},
            {'[', ']'},
            {'{', '}'} 
        };

        Stack<char> parentheses_stack = new Stack<char>();

        foreach (char letter in s) {
            if (dict_keys.ContainsKey(letter)) {
                parentheses_stack.Push(letter);
            } else {
                if (parentheses_stack.Count == 0) {
                    return false;
                } else {
                    if (dict_keys[parentheses_stack.Pop()] != letter) {
                        return false;
                    }
                }
            }
        }

        if (parentheses_stack.Count > 0) {
            return false;
        }

        return true;

    }
}
